File: ios/LaneShadow/Views/Molecules/HighlightTagsStagger.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Highlight Tag Model

/**
 * Highlight tag model
 *
 * Represents a single highlight tag with optional emoji icon.
 */
public struct LSHighlightTag: Sendable, Identifiable, Equatable {
    public let id: String
    public let label: String
    public let icon: String?

    public init(
        id: String = UUID().uuidString,
        label: String,
        icon: String? = nil
    ) {
        self.id = id
        self.label = label
        self.icon = icon
    }
}

// MARK: - Highlight Tags Stagger Component

/**
 * Highlight tags stagger molecule component
 *
 * Staggered fade-in animation for highlight tags in a wrapping layout.
 * Following React Native component from react-native/components/enrichment/highlight-tags-stagger.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Tag background: `theme.colors.primary.default` at 10% opacity
 *   - Tag border: `theme.colors.primary.default` at 30% opacity (1pt)
 *   - Tag text: `theme.colors.primary.default`
 * - Layout:
 *   - Container gap: 8pt
 *   - Tag padding: horizontal `theme.space.md`, vertical `theme.space.xs`
 *   - Tag corner radius: `theme.radius.full` (capsule/pill)
 *   - Tag icon gap: 4pt
 * - Typography:
 *   - Tag text: `theme.type.label.md`
 *   - Tag icon: 14pt
 *
 * ## Behavior
 * - Staggered fade-in animation with configurable delay (default 100ms)
 * - Scale animation from 0.95 to 1.0 (subtle pop)
 * - Shows EmptyView when not visible or highlights array is empty
 * - Respects reduce-motion preference (instant reveal when enabled)
 * - Tags wrap in FlowLayout when space runs out
 *
 * ## Parameters
 * - highlights: Array of highlight tags to display
 * - visible: Whether the tags are ready to show
 * - staggerDelay: Stagger delay between each tag in ms (default 100)
 * - fadeDuration: Fade-in duration per tag in ms (default 300)
 * - scaleDuration: Scale pop animation duration in ms (default 300)
 * - testID: Optional testing identifier for UI tests
 *
 * ## Accessibility
 * - Container announces "{count} route highlights"
 * - Each tag has accessibility label
 */
public struct LSHighlightTagsStagger: View {
    @Environment(\.theme) private var theme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    @State private var tagsVisible: [Bool] = []

    private let highlights: [LSHighlightTag]
    private let visible: Bool
    private let staggerDelay: Int
    private let fadeDuration: Int
    private let scaleDuration: Int
    private let testID: String

    public init(
        highlights: [LSHighlightTag],
        visible: Bool,
        staggerDelay: Int = 100,
        fadeDuration: Int = 300,
        scaleDuration: Int = 300,
        testID: String = "highlight-tags"
    ) {
        self.highlights = highlights
        self.visible = visible
        self.staggerDelay = staggerDelay
        self.fadeDuration = fadeDuration
        self.scaleDuration = scaleDuration
        self.testID = testID
        _tagsVisible = State(initialValue: Array(repeating: false, count: highlights.count))
    }

    // MARK: - Body

    public var body: some View {
        Group {
            if visible, !highlights.isEmpty {
                FlowLayout(spacing: 8) {
                    ForEach(Array(highlights.enumerated()), id: \.element.id) { index, tag in
                        TagChip(
                            tag: tag,
                            isVisible: tagsVisible.indices.contains(index) ? tagsVisible[index] : false,
                            delay: index * staggerDelay,
                            fadeDuration: fadeDuration,
                            scaleDuration: scaleDuration,
                            reduceMotion: reduceMotion,
                            testID: "\(testID)-\(index)"
                        )
                    }
                }
                .accessibilityElement(children: .contain)
                .accessibilityLabel("\(highlights.count) route highlights")
                .accessibilityIdentifier(testID)
                .onAppear {
                    showTags()
                }
                .onChange(of: visible) { _, newValue in
                    if newValue {
                        showTags()
                    } else {
                        hideTags()
                    }
                }
            }
        }
    }

    // MARK: - Helpers

    private func showTags() {
        if reduceMotion {
            // Instant reveal for reduce motion
            tagsVisible = Array(repeating: true, count: highlights.count)
        } else {
            // Staggered animation
            for index in 0 ..< highlights.count {
                DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(index * staggerDelay)) {
                    if index < tagsVisible.count {
                        tagsVisible[index] = true
                    }
                }
            }
        }
    }

    private func hideTags() {
        tagsVisible = Array(repeating: false, count: highlights.count)
    }
}

// MARK: - Tag Chip View

private struct TagChip: View {
    @Environment(\.theme) private var theme

    let tag: LSHighlightTag
    let isVisible: Bool
    let delay: Int
    let fadeDuration: Int
    let scaleDuration: Int
    let reduceMotion: Bool
    let testID: String

    @State private var scale: CGFloat = 0.95

    var body: some View {
        HStack(spacing: 4) {
            if let icon = tag.icon {
                Text(icon)
                    .font(.system(size: 14))
            }

            Text(tag.label)
                .font(theme.type.label.md.font)
        }
        .foregroundStyle(tagTextColor)
        .padding(.horizontal, theme.space.md)
        .padding(.vertical, theme.space.xs)
        .background(tagBackground)
        .overlay(
            Capsule()
                .stroke(tagBorderColor, lineWidth: 1)
        )
        .clipShape(Capsule())
        .scaleEffect(scale)
        .opacity(isVisible ? 1 : 0)
        .animation(.easeOut(duration: reduceMotion ? 0 : Double(fadeDuration) / 1000), value: isVisible)
        .onAppear {
            if !reduceMotion {
                DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(delay)) {
                    withAnimation(.easeOut(duration: Double(scaleDuration) / 1000)) {
                        scale = 1.0
                    }
                }
            } else {
                scale = 1.0
            }
        }
        .accessibilityLabel(tag.label)
        .accessibilityIdentifier(testID)
    }

    private var tagTextColor: Color {
        theme.colors.primary.default
    }

    private var tagBackground: some View {
        theme.colors.primary.default.opacity(0.1)
    }

    private var tagBorderColor: Color {
        theme.colors.primary.default.opacity(0.3)
    }
}

// MARK: - Flow Layout (Vertical Wrapping)

/**
 * Simple flow layout for wrapping tags
 *
 * Arranges children in rows, wrapping to the next line when space runs out.
 * Reused from SuggestionChips.swift to maintain consistency.
 */
private struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let result = FlowResult(
            in: proposal.replacingUnspecifiedDimensions().width,
            subviews: subviews,
            spacing: spacing
        )
        return result.size
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        let result = FlowResult(
            in: bounds.width,
            subviews: subviews,
            spacing: spacing
        )
        for (index, subview) in subviews.enumerated() {
            subview.place(
                at: CGPoint(x: bounds.minX + result.positions[index].x, y: bounds.minY + result.positions[index].y),
                proposal: .unspecified
            )
        }
    }

    struct FlowResult {
        var size: CGSize
        var positions: [CGPoint]

        init(in maxWidth: CGFloat, subviews: Subviews, spacing: CGFloat) {
            var positions: [CGPoint] = []
            var currentX: CGFloat = 0
            var currentY: CGFloat = 0
            var rowHeight: CGFloat = 0

            for subview in subviews {
                let size = subview.sizeThatFits(.unspecified)

                if currentX + size.width > maxWidth, currentX > 0 {
                    // Move to next row
                    currentX = 0
                    currentY += rowHeight + spacing
                    rowHeight = 0
                }

                positions.append(CGPoint(x: currentX, y: currentY))
                currentX += size.width + spacing
                rowHeight = max(rowHeight, size.height)
            }

            size = CGSize(width: maxWidth, height: currentY + rowHeight)
            self.positions = positions
        }
    }
}

// MARK: - Preview

#Preview("HighlightTagsStagger - Visible") {
    VStack(alignment: .leading, spacing: 16) {
        Text("Route Highlights")
            .font(.headline)
            .padding(.horizontal)

        LSHighlightTagsStagger(
            highlights: [
                LSHighlightTag(label: "Scenic", icon: "🛣️"),
                LSHighlightTag(label: "Coastal", icon: "🌊"),
                LSHighlightTag(label: "Mountains", icon: "⛰️"),
                LSHighlightTag(label: "Wine country", icon: "🍇"),
                LSHighlightTag(label: "Historic"),
                LSHighlightTag(label: "Day trip"),
            ],
            visible: true
        )

        Spacer()
    }
    .laneShadowTheme()
}

#Preview("HighlightTagsStagger - Skeleton") {
    VStack(alignment: .leading, spacing: 16) {
        Text("Route Highlights")
            .font(.headline)
            .padding(.horizontal)

        LSHighlightTagsStagger(
            highlights: [
                LSHighlightTag(label: "Scenic", icon: "🛣️"),
                LSHighlightTag(label: "Coastal", icon: "🌊"),
            ],
            visible: false
        )

        Spacer()
    }
    .laneShadowTheme()
}

#Preview("HighlightTagsStagger - Empty") {
    VStack(alignment: .leading, spacing: 16) {
        Text("Route Highlights")
            .font(.headline)
            .padding(.horizontal)

        LSHighlightTagsStagger(
            highlights: [],
            visible: true
        )

        Spacer()
    }
    .laneShadowTheme()
}

#Preview("HighlightTagsStagger - Transition") {
    struct TransitionDemo: View {
        @State private var visible = false

        var body: some View {
            VStack(alignment: .leading, spacing: 24) {
                LSHighlightTagsStagger(
                    highlights: [
                        LSHighlightTag(label: "Scenic", icon: "🛣️"),
                        LSHighlightTag(label: "Coastal", icon: "🌊"),
                        LSHighlightTag(label: "Mountains", icon: "⛰️"),
                        LSHighlightTag(label: "Wine country", icon: "🍇"),
                    ],
                    visible: visible
                )

                Button("Toggle Visibility") {
                    withAnimation(.easeInOut(duration: 0.3)) {
                        visible.toggle()
                    }
                }
                .buttonStyle(.bordered)
                .padding(.horizontal)
            }
            .padding(.vertical)
            .laneShadowTheme()
        }
    }

    return TransitionDemo()
}

#Preview("HighlightTagsStagger - Custom Timing") {
    VStack(alignment: .leading, spacing: 16) {
        Text("Custom Stagger (200ms)")
            .font(.headline)
            .padding(.horizontal)

        LSHighlightTagsStagger(
            highlights: [
                LSHighlightTag(label: "Scenic", icon: "🛣️"),
                LSHighlightTag(label: "Coastal", icon: "🌊"),
                LSHighlightTag(label: "Mountains", icon: "⛰️"),
            ],
            visible: true,
            staggerDelay: 200,
            fadeDuration: 500,
            scaleDuration: 400
        )

        Spacer()
    }
    .laneShadowTheme()
}

#Preview("HighlightTagsStagger - Reduce Motion") {
    VStack(alignment: .leading, spacing: 16) {
        Text("Reduce Motion")
            .font(.headline)
            .padding(.horizontal)

        LSHighlightTagsStagger(
            highlights: [
                LSHighlightTag(label: "Scenic", icon: "🛣️"),
                LSHighlightTag(label: "Coastal", icon: "🌊"),
                LSHighlightTag(label: "Mountains", icon: "⛰️"),
            ],
            visible: true
        )

        Spacer()
    }
    .laneShadowTheme()
    .accessibilityReduceMotion(true)
}

#Preview("HighlightTagsStagger - Without Icons") {
    VStack(alignment: .leading, spacing: 16) {
        Text("Without Icons")
            .font(.headline)
            .padding(.horizontal)

        LSHighlightTagsStagger(
            highlights: [
                LSHighlightTag(label: "Scenic"),
                LSHighlightTag(label: "Coastal"),
                LSHighlightTag(label: "Mountains"),
                LSHighlightTag(label: "Wine country"),
            ],
            visible: true
        )

        Spacer()
    }
    .laneShadowTheme()
}

#Preview("HighlightTagsStagger - Complete") {
    VStack(alignment: .leading, spacing: 24) {
        Text("Highlight Tags Stagger Showcase")
            .font(.title)
            .fontWeight(.bold)
            .padding(.horizontal)

        VStack(alignment: .leading, spacing: 8) {
            Text("With Icons")
                .font(.headline)
                .padding(.horizontal)

            LSHighlightTagsStagger(
                highlights: [
                    LSHighlightTag(label: "Scenic", icon: "🛣️"),
                    LSHighlightTag(label: "Coastal", icon: "🌊"),
                    LSHighlightTag(label: "Mountains", icon: "⛰️"),
                    LSHighlightTag(label: "Wine country", icon: "🍇"),
                ],
                visible: true
            )
        }

        VStack(alignment: .leading, spacing: 8) {
            Text("Without Icons")
                .font(.headline)
                .padding(.horizontal)

            LSHighlightTagsStagger(
                highlights: [
                    LSHighlightTag(label: "Scenic"),
                    LSHighlightTag(label: "Coastal"),
                    LSHighlightTag(label: "Mountains"),
                ],
                visible: true
            )
        }

        VStack(alignment: .leading, spacing: 8) {
            Text("Mixed")
                .font(.headline)
                .padding(.horizontal)

            LSHighlightTagsStagger(
                highlights: [
                    LSHighlightTag(label: "Scenic", icon: "🛣️"),
                    LSHighlightTag(label: "Coastal"),
                    LSHighlightTag(label: "Mountains", icon: "⛰️"),
                ],
                visible: true
            )
        }

        Spacer()
    }
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/HighlightTagsStagger.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalInspectionMode
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Highlight tag data model
 *
 * @param label Display text for the tag
 * @param icon Optional emoji or icon character
 */
data class HighlightTag(
    val label: String,
    val icon: String? = null,
)

/**
 * HighlightTagsStagger molecule component
 *
 * Staggered fade-in animation for highlight tags.
 * Each tag fades in with a configurable delay (default 100ms).
 * Tags use a chip/pill style with optional emoji icons.
 * Wraps to multiple lines using flex-wrap.
 *
 * ## Design Tokens
 * - Background: primary.default at 10% opacity
 * - Border: primary.default at 30% opacity
 * - Text: label.md typography, primary.default color
 * - Radius: full (pill shape)
 * - Padding: horizontal md, vertical xs
 * - Icon size: 14sp
 * - Gap between tags: 8dp (space.sm)
 *
 * ## Animation
 * - Fade-in: 0 to 1 alpha over fadeDuration (default 300ms)
 * - Scale: 0.95 to 1.0 over scaleDuration (default 300ms)
 * - Stagger delay between tags: staggerDelay (default 100ms)
 * - Respects reduce-motion preference (instant reveal when enabled)
 *
 * ## Accessibility
 * - Container has accessibility label describing the count
 * - Each tag has accessibility label matching its text
 *
 * @param highlights Array of highlight tags to display
 * @param visible Whether the tags are ready to show
 * @param staggerDelay Stagger delay between each tag in ms (default 100)
 * @param fadeDuration Fade-in duration per tag in ms (default 300)
 * @param scaleDuration Scale pop animation duration in ms (default 300)
 * @param modifier Modifier for the container
 * @param testID Test ID for UI testing
 */
@Composable
fun HighlightTagsStagger(
    highlights: List<HighlightTag>,
    visible: Boolean,
    staggerDelay: Int = 100,
    fadeDuration: Int = 300,
    scaleDuration: Int = 300,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current
    val isInInspectionMode = LocalInspectionMode.current

    // Return nothing if not visible or empty
    if (!visible || highlights.isEmpty()) {
        return
    }

    // Build accessibility description for container
    val accessibilityDescription = "${highlights.size} route highlights"

    // Container with flex wrap using Row with horizontalArrangement.spacedBy
    Row(
        modifier = modifier
            .testTag(testID ?: "highlight-tags")
            .semantics {
                contentDescription = accessibilityDescription
            },
        horizontalArrangement = Arrangement.spacedBy(theme.space.sm), // 8dp gap
    ) {
        highlights.forEachIndexed { index, tag ->
            HighlightTagItem(
                tag = tag,
                delay = index * staggerDelay,
                fadeDuration = fadeDuration,
                scaleDuration = scaleDuration,
                isInInspectionMode = isInInspectionMode,
                theme = theme,
                testTag = testID?.let { "$it-$index" } ?: "highlight-tags-$index",
            )
        }
    }
}

/**
 * Individual highlight tag item with entrance animation
 *
 * @param tag The highlight tag to display
 * @param delay Animation delay in ms
 * @param fadeDuration Fade-in duration in ms
 * @param scaleDuration Scale animation duration in ms
 * @param isInInspectionMode Whether in preview/inspection mode
 * @param theme LaneShadow theme values
 * @param testTag Test tag for UI testing
 */
@Composable
private fun HighlightTagItem(
    tag: HighlightTag,
    delay: Int,
    fadeDuration: Int,
    scaleDuration: Int,
    isInInspectionMode: Boolean,
    theme: com.laneshadow.theme.LaneShadowThemeValues,
    testTag: String,
) {
    // Animation states
    var alpha by remember { mutableFloatStateOf(0f) }
    var scale by remember { mutableFloatStateOf(0.95f) }

    // Trigger animations when component is first composed
    LaunchedEffect(Unit) {
        if (isInInspectionMode) {
            // Skip animation in inspection mode (previews, tests)
            alpha = 1f
            scale = 1f
        } else {
            // Animate fade-in
            alpha = 1f

            // Animate scale with delay
            kotlinx.coroutines.delay(delay.toLong())
            scale = 1f
        }
    }

    // Background color: primary.default at 10% opacity
    val backgroundColor = theme.colors.primary.default.copy(alpha = 0.1f)

    // Border color: primary.default at 30% opacity
    val borderColor = theme.colors.primary.default.copy(alpha = 0.3f)

    // Text color: primary.default
    val textColor = theme.colors.primary.default

    // Icon and text style: 14sp for icon, label.md for text
    val iconStyle = TextStyle(fontSize = 14.sp)
    val textStyle = theme.type.label.md

    Surface(
        modifier = Modifier
            .alpha(alpha)
            .scale(scale)
            .testTag(testTag)
            .semantics {
                contentDescription = tag.label
            },
        shape = theme.radius.full.toRoundedCornerShape(), // Pill shape
        color = backgroundColor,
        border = BorderStroke(1.dp, borderColor),
    ) {
        Row(
            modifier = Modifier.padding(
                horizontal = theme.space.md, // 16dp horizontal padding
                vertical = theme.space.xs,  // 4dp vertical padding
            ),
            horizontalArrangement = Arrangement.spacedBy(theme.space.xs), // 4dp gap between icon and text
        ) {
            // Optional icon
            if (tag.icon != null) {
                Text(
                    text = tag.icon,
                    style = iconStyle,
                )
            }

            // Label text
            Text(
                text = tag.label,
                style = textStyle,
                color = textColor,
            )
        }
    }
}

/**
 * Extension to convert Dp radius to RoundedCornerShape
 */
private fun Dp.toRoundedCornerShape() = androidx.compose.foundation.shape.RoundedCornerShape(this)
